compiler/test/dotty/tools/vulpix/ParallelTesting.scala

```diff
@@ -178,12 +179,80 @@ trait ParallelTesting extends RunnerOrchestration { self =>
         .getOrElse("")
       }
       .toList.sortBy(_._1).map(_._2.filter(isSourceFile).sorted)
+
+    def sourceFiles: Array[JFile] = compilationGroups.flatten.filter(isSourceFile).toArray
+  }
+
+  private trait CompilationLogic { this: Test =>
+    val suppressErrors = false
+
+    final def compileTestSource(testSource: TestSource): Either[Throwable, List[TestReporter]] =
+      Try(testSource match {
+        case testSource @ JointCompilationSource(name, files, flags, outDir, fromTasty, decompilation) =>
+          val reporter =
+            if (fromTasty) compileFromTasty(               flags, suppressErrors, outDir)
+            else           compile(testSource.sourceFiles, flags, suppressErrors, outDir)
+          List(reporter)
+
+        case testSource @ SeparateCompilationSource(_, dir, flags, outDir) =>
+          testSource.compilationGroups.map(files => compile(files, flags, suppressErrors, outDir))  // TODO? only `compile` option?
+      }).toEither
+
+    final def countErrorsAndWarnings(reporters: Seq[TestReporter]): (Int, Int) =
+      reporters.foldLeft((0, 0)) { case ((err, warn), r) => (err + r.errorCount, warn + r.warningCount) }
+
+    final def countErrors  (reporters: Seq[TestReporter]) = countErrorsAndWarnings(reporters)._1
+    final def countWarnings(reporters: Seq[TestReporter]) = countErrorsAndWarnings(reporters)._2
+    final def reporterFailed(r: TestReporter) = r.compilerCrashed || r.errorCount > 0
+
+    final def checkFile(testSource: TestSource): Option[JFile] = (testSource match {
+      case ts: JointCompilationSource =>
+        ts.files.filter(f => !f.isDirectory).map { f => new JFile(f.getAbsolutePath.replaceFirst("\\.scala$", ".check")) }.headOption
+
+      case ts: SeparateCompilationSource =>
+        Option(new JFile(ts.dir.getAbsolutePath + ".check"))
+    }).filter(_.exists)
+
+    final def diffTest(testSource: TestSource, checkFile: JFile, actual: List[String]) = {
+      val expected = Source.fromFile(checkFile, "UTF-8").getLines().toList
+      for (msg <- diffMessage(testSource.title, actual, expected)) {
+        echo(msg)
+        failTestSource(testSource)
+        dumpOutputToFile(checkFile, actual)
+      }
+    }
+
+    final def encapsulatedCompilation(testSource: TestSource) = new LoggedRunnable { self =>
+      def checkTestSource(): Unit = tryCompile(testSource) {
+        val reportersOrCrash = compileTestSource(testSource)
+        onComplete(testSource, reportersOrCrash, self)
+        registerCompletion()
+      }
+    }
+
+    final def onComplete(testSource: TestSource, reportersOrCrash: Either[Throwable, Seq[TestReporter]], logger: LoggedRunnable): Unit =
```

Comment:
IMO `fold` is more elegant and idiomatic. In general, I prefer functional control flow abstractions to general-purpose ones, since the former ones are the tool meant specifically for the effect types.

Reply:
The argument may apply to the people new to Scala. But there's a clear advantage in using the idiomatic style: reduction of LOC number and predictability of the behavior.

This is exactly why `goto` is discouraged in favour of `if`, `for` etc, and this is exactly why Scala has dozens of specialized, well-behaved, high-level collection methods to substitute the low-level, though not as numerous, Java API.

Reply:
I would follow @allanrenucci opinion here. Pattern matching will be a bit more verbose but the sopes of the code will be clearer.